tests/functional/conftest.py

```diff
@@ -48,10 +51,40 @@
 # Directory storing test-specific data.
 _DATA_DIR = os.path.join(os.path.dirname(os.path.abspath(__file__)), 'data')
 
-# Provide test access to _DATA_DIR via a fixture.
+# Code
+# ====
+# Fixtures
+# --------
+# This class provides access to the _DATA_DIR, and also copies correctly-named
+# directories in the _DATA_DIR to the tmpdir.
+class DataDir(object):
+    # Optionally copy from _DATA_DIR to tmpdir.
+    def __init__(self,
+      # The request object for this test. See
+      # https://pytest.org/latest/builtin.html#_pytest.python.FixtureRequest
+      # and
+      # https://pytest.org/latest/fixture.html#fixtures-can-introspect-the-requesting-test-context.
+      request,
+      # The tmpdir object for this test. See
+      # https://pytest.org/latest/tmpdir.html.
+      tmpdir):
+
+        # Provide access to _DATA_DIR, using the same nomenclature as tmpdir.
+        self.strpath = _DATA_DIR
+
+        # Strip the leading "test_' from the test's name.
+        name = request.function.__name__[5:]
+        # Copy _DATA_DIR/<name> if it exists to the tmpdir.
+        source_data = os.path.join(_DATA_DIR, name)
+        if os.path.exists(source_data):
+            # Provide access to the location to which the directory was copied.
+            self.dest_strpath = os.path.join(tmpdir.strpath, 'data', name)
+            shutil.copytree(source_data, self.dest_strpath)
+
+# Define a fixure for the DataDir object.
 @pytest.fixture
-def data_dir():
-    return _DATA_DIR
+def data_dir(request, tmpdir):
+    return DataDir(request, tmpdir)
```

Comment:
Thus, all of the code in `DataDir.__init__()` should be moved into `data_dir()`, and slightly modified to return a `py.path.local` object instead of a `DataDir` object.

It's fair to raise an error if a test requests a `data_dir` but does not have a corresponding folder in `tests/functional/data`